components/concept/push/src/main/java/mozilla/components/concept/push/PushProcessor.kt

```diff
@@ -6,74 +6,95 @@
      */
-    @CallSuper
-    open fun initialize() {
-        instance = this
-    }
-
-    /**
-     * Start the push processor and starts any service associated.
-     */
-    abstract fun start()
+    fun initialize()
 
     /**
-     * Stop the push service and stops any service associated.
+     * Removes all push subscriptions from the device.
      */
-    abstract fun stop()
+    fun shutdown()
 
     /**
      * A new registration token has been received.
      */
-    abstract fun onNewToken(newToken: String)
+    fun onNewToken(newToken: String)
 
     /**
      * A new push message has been received.
      */
-    abstract fun onMessageReceived(message: PushMessage)
+    fun onMessageReceived(message: EncryptedPushMessage)
 
     /**
      * An error has occurred.
      */
-    abstract fun onError(error: Error)
+    fun onError(error: PushError)
 
     companion object {
+        /**
+         * Initialize the PushProcessor that needs to be called before
+         */
+        fun install(processor: PushProcessor) {
+            instance = processor
+        }
+
         @Volatile
-        var instance: PushProcessor? = null
+        private var instance: PushProcessor? = null
+
+        @VisibleForTesting
+        internal fun reset() {
+            instance = null
+        }
         val requireInstance: PushProcessor
             get() = instance ?: throw IllegalStateException(
-                "You need to call start() on your Push instance from Application.onCreate()."
+                "You need to call initialize() on your Push instance from Application.onCreate()."
             )
     }
 }
 
 /**
  * A push message holds the information needed to pass the message on to the appropriate receiver.
  */
-data class PushMessage(val data: String)
-
-/**
- * The different kinds of push messages.
- */
-enum class PushType {
-    Services,
-    ThirdParty
+data class EncryptedPushMessage(
+    val channelId: String,
+    val body: String,
+    val encoding: String,
+    val salt: String = "",
+    val cryptoKey: String = "" // diffie–hellman key
+) {
+    companion object {
+        /**
+         * The [salt] and [cryptoKey] are optional as part of the standard for WebPush, so we should default
+         * to empty strings.
+         *
+         * Note: We have use the invoke operator since secondary constructors don't work with nullable primitive types.
+         */
+        operator fun invoke(
+            channelId: String,
+            body: String,
+            encoding: String,
+            salt: String? = null,
+            cryptoKey: String? = null
+        ) = EncryptedPushMessage(channelId, body, encoding, salt ?: "", cryptoKey ?: "")
+    }
 }
 
 /**
  *  Various error types.
  */
-sealed class Error {
-    data class RegistrationError(val desc: String) : Error()
-    data class NetworkError(val desc: String) : Error()
+sealed class PushError(open val desc: String) {
```

Comment:
nit: you could just define `desc` on member classes, without the overrides

Reply:
If you have a when expression, you would need to cast it to a specific type to access the description then (see AutoPushFeature#onError which is one use case).